DirectXHost.cs

```diff
@@ -8,230 +8,261 @@
+        public event EventHandler HostWindowCreated;
+        protected void OnHostWindowCreated()
+        {
+            if (HostWindowCreated != null)
+            {
+                HostWindowCreated(this._dxForm, new EventArgs());
+            }
+        }
+
+        protected void OnHostWindowClosed()
+        {
+            if (HostWindowClosed != null)
+            {
+                HostWindowClosed(this._dxForm, new EventArgs());
+            }
+        }
+        #endregion
+
+
+        public void Exit()
+        {
+            throw new NotImplementedException();
+        }
+
+        public async Task Run()
+        {
+            try
+            {
+                await Settings.Load();
+                InitializeRenderForm();
+
+                _graphics = new GraphicsD3D11();
+                _graphics.Initialise(_dxForm, true);
+                _dxForm.UserResized += (sender, args) =>
+                {
+                    var renderForm = sender as RenderForm;
+                    ClientSize = new Size(renderForm.ClientSize.Width, renderForm.ClientSize.Height);
+                    UserResized = true;
+                };
+                _overlayForm.Show();
+                RenderLoop.Run(_dxForm, RenderCallback, true);
+            }
+            catch (ArgumentException ex)
+            {
+                MessageBox.Show(string.Format("Exception running DirectX host window.\r\n{0}", ex.Message), "DirectX Host", MessageBoxButtons.OK, MessageBoxIcon.Error);
+            }
+            catch (Exception ex)
+            {
+                MessageBox.Show(string.Format("Exception running DirectX host window.\r\n{0}", ex.Message), "DirectX Host", MessageBoxButtons.OK, MessageBoxIcon.Error);
+            }
+        }
+
+        // Set the bitmap object to the size of the screen
+        private Bitmap _bmpScreenshot;
+        Bitmap bmpScreenshot
+        {
+            get => _bmpScreenshot;
+            set
+            {
+                _bmpScreenshot = value;
+                _overlayTarget.Image = value;
+                updateImageBounds();
+            }
+        }
+
+        Graphics gfxScreenshot;
+        private void RenderCallback()
+        {
+            if (_dxForm.WindowState != FormWindowState.Minimized)
+            {
+                Update();
+                Draw();
+                gfxScreenshot = Graphics.FromImage(bmpScreenshot);
+                IntPtr dc = gfxScreenshot.GetHdc();
+                PrintWindow(_dxForm.Handle, dc, 0);
+                gfxScreenshot.ReleaseHdc();
+                gfxScreenshot.Dispose();
+                _overlayTarget.Invalidate();
+            }
+            // Calculate Frame Limiting
+            //long end = DateTime.Now.Ticks;
+            //long duration = end - time;
+            int framerate = Settings.frameRate;
+            if (Settings.frameRate > 0)
+                Thread.Sleep(new TimeSpan(10000000 / Settings.frameRate));

```

Comment:
I agree the original was missing the frame limit reference, which was accidental. But the proposed change is not an accurate frame limit as it does not take into account the actual duration of the frame. Should be the following:
```C#
if (Settings.frameRate > 0) {
	long end = DateTime.Now.Ticks;
	long duration = end - time;
	int timeout = (1000 / Settings.frameRate) - (int)new TimeSpan(duration).TotalMilliseconds;
	if (timeout > 0)
		Thread.Sleep(timeout);
}
```

Reply:
New proposed change uses Stopwatch which is more accurate.